src/NServiceBus.Transport.Msmq/MessagePump.cs

```diff
@@ -157,43 +159,44 @@ async Task InnerProcessMessages()
                         continue;
                     }
 
-                    if (cancellationTokenSource.IsCancellationRequested)
+                    if (messagePumpCancellationTokenSource.IsCancellationRequested)
                     {
                         return;
                     }
 
-                    await concurrencyLimiter.WaitAsync(cancellationToken).ConfigureAwait(false);
+                    await concurrencyLimiter.WaitAsync(messagePumpCancellationTokenSource.Token).ConfigureAwait(false);
 
-                    _ = ReceiveMessage();
+                    _ = Task.Factory.StartNew(state => InnerReceiveMessages((MessagePump)state),
+                        this,  // We pass a state to make sure we benefit from lamda delegate caching. See https://github.com/Particular/NServiceBus/issues/3884
+                        CancellationToken.None,  // CancellationToken.None is used here since cancelling the task before it can run can cause the concurrencyLimiter to not be released
+                        TaskCreationOptions.DenyChildAttach,
+                        TaskScheduler.Default)
+                        .Unwrap();
```

Comment:
Strictly speaking, you wouldn't have to. My brain would stumble over it though because async body in `Task.Factory.StartNew` is always a dangerous territory. I would always prefer here the safe by default approach even though your point is absolutely valid

Reply:
I'm fine either way. It seems the former (and now inlined) helper method correctly used `Unwrap` but in this specific case it wasn't used, so we could optimize it away. Anyway, no big deal.